### PR TITLE
feat: Resolve service.name based on Kubernetes attributes

### PR DESCRIPTION
**Pull request by skhalash**
## Description

1. New feature: Resolve service.name Resource Attribute
This feature works in three main steps:

* Label Extraction: Extract the values of the app and app.kubernetes.io/name labels into resource attributes.
* Attribute Resolution: Determine the service.name either based on attributes extracted in step 1 or existing k8s attributes.
* Cleanup: Remove temporary attributes used during the extraction and resolution steps.
Introduction of Common Gateway Processors Package

2. As the feature sets for both trace and metric gateways continue to grow, this PR introduces a new package that contains common gateway processors. This aims to reduce code duplication and simplify maintenance.

3. The PR also moves existing gateway configuration marshaling tests to use golden files, aiming for clearer and more maintainable test cases.

This refined PR description should offer reviewers a more structured and detailed overview of the changes, making it easier to understand the impact and reason behind them.

## Related Issues and Documents

Related issues:
https://github.com/kyma-project/telemetry-manager/issues/373

## Traceability

- [x] The PR is linked to a GitHub Issue.
- [x] New features have a milestone label set.
- [x] New features have defined acceptance criteria in a corresponding GitHub Issue, and all criteria are satisfied with this PR.
- [x] The corresponding GitHub Issue has a respective `area` label.
- [ ] The follow-up issues (if any) are linked in the `Related Issues` section.

## Testability

The feature is unit-tested:

- [x] Yes.
- [ ] No, because unit tests are not needed.
- [ ] No, because of ...

The feature is e2e-tested:

- [x] Yes.
- [ ] No, because e2e-tests are not needed.
- [ ] No, because of ...

Tests conducted for the PR:

## Codebase

- [x] My code follows the [Effective Go](https://go.dev/doc/effective_go) style guidelines.
- [x] The code was planned and designed following the defined architecture and the separation of concerns.
- [x] The code has sufficient comments, particularly for all hard-to-understand areas.
- [x] This PR adds value and shows no feature creep.
- [x] I have augmented the test suite that proves my fix is effective or that my feature works.
- [x] Adjusted the documentation if the change is user-facing.

### CODE REVIEW

File: docs/user/README.md [hunk 1]
```diff
@@ -23,6 +23,7 @@ To support telemetry for your applications, Kyma's Telemetry module provides the
 - Tooling for collection, filtering, and shipment: Based on the [Open Telemetry Collector](https://opentelemetry.io/docs/collector/) and [Fluent Bit](https://fluentbit.io/), you can configure basic pipelines to filter and ship telemetry data.
 - Integration in a vendor-neutral way to a vendor-specific observability system (traces and metrics only): Based on the [OpenTelemetry protocol (OTLP)](https://opentelemetry.io/docs/reference/specification/protocol/), you can integrate backend systems.
 - Guidance for the instrumentation (traces and metrics only): Based on [Open Telemetry](https://opentelemetry.io/), you get community samples on how to instrument your code using the [Open Telemetry SDKs](https://opentelemetry.io/docs/instrumentation/) in nearly every programming language.
+- Enriching telemetry data by adding common attributes (traces and metrics only). This is done in compliance with established semantic conventions, ensuring that the enriched data adheres to industry best practices and is more meaningful for analysis.
```
Comment (NHingerl) [inline docs/user/README.md:26]:
```suggestion
- [Enriching](#automatic-telemetry-enrichment) telemetry data by automatically adding common attributes (traces and metrics only). This is done in compliance with established semantic conventions, ensuring that the enriched data adheres to industry best practices and is more meaningful for analysis.
```

We can link to that section.
And I'm not sure whether it was by purpose that you didn't mention that it's automatic? If by purpose, drop the "automatically" from my suggestion :)

File: docs/user/README.md [hunk 2]
```diff
@@ -65,6 +66,22 @@ The metric gateway and agent are based on an [OTel Collector](https://openteleme
...
+
+Kyma's Telemetry module automatically enriches your data by adding the following attributes (traces and metrics only):
+
+- `service.name`: The logical name of the service that emits the telemetry data. If not provided by the user, it is populated from Kubernetes metadata, based on the following hierarchy of labels and names:
+  1. `app.kubernetes.io/name` Pod label value.
+  2. `app` Pod label value.
+  3. Deployment/DaemonSet/StatefulSet/Job name.
+  4. Pod name.
+  5. If none of the above is available, the value is `unknown_service`.
+- `k8s.*` attributes: These attributes encapsulate various pieces of Kubernetes metadata associated with the pod, including but not limited to:
```
Comment (NHingerl) [inline docs/user/README.md:79]:
```suggestion
- `k8s.*` attributes: These attributes encapsulate various pieces of Kubernetes metadata associated with the Pod, including but not limited to:
```

File: docs/user/README.md (hunk 1 above)
Comment (skhalash) [reply docs/user/README.md]: It is indeed automatic, the user does not have to do anything to get telemetry data enriched

**Review by NHingerl — APPROVED** [review]

**Review by NHingerl — APPROVED** [review]

**Review by rakesh-garimella — APPROVED** [review]

**Pull request merged**